### ISSUE TITLE
glslify will not import (via "require" or "import")

### ISSUE THREAD

**Issue by Johnrobmiller** [glslify/glslify#143]
glslify looks great, but I think I might have to go without using glslify until I can solve whatever is going on here.

glslify will not import.  Here is what I tried:

- const glsl = require('glslify')
- import { glsl } from 'glslify'
- import glsl from 'glslify'
- import * as glsl from 'glslify
- import { createRequire } from 'module'; const require = createRequire(import.meta.url); const glsl = require('glslify')
- installing the glslify babel-plugin
- reinstalling commonjs
- installing glslify with and without using th -g tag

No matter what I do, glslify will not import.  I get the same errors every time.  It's either:
- Could not find a declaration file for module 'glslify' (when using import)
- reference error: require is not defined (when using require)

I don't think my setup is too much different from anybody else's.  It's just a snowpack build setup with React and Typescript.

Perhaps there was some update in babel or snowpack that broke glslify.  I have no idea.  Either way, I'm submitting this issue so that it can be out in the public eye, and if I am lucky someone will be able to help me out and post a solution.

**Comment by homerjam** [comment]
I'm having similar issues I think - pretty sure it's down to [this](https://github.com/glslify/glslify/blob/ba9c52c46e10068215290753e99ac516cc02d0d7/package.json#L65). Some renaming would at least allow to explicitly require the main module file when required. Right now it's completely impossible.

**Comment by Johnrobmiller** [comment]
Thanks homerjam.  I'm working on something else all day today, but I'll get this a whirl tomorrow.

**Comment by Johnrobmiller** [comment]
Welp, I tried the fix above, but sadly, it looks like I've got bigger problems.  I wasn't able to modify glslify's package.json because glslify never actually installed.  I checked my node_modules, and it is not there.  Still, npm reported to have installed it without error.  Very strange.

**Comment by jeffpeck10x** [comment]
Is there a fix for this? Or has anyone found a suitable workaround?

**Comment by jeffpeck10x** [comment]
Oh, this works: https://github.com/onnovisser/babel-plugin-glsl

**Comment by Allakazan** [comment]
Anyone has ANY fix for this ????? I can't use glslify with typescript and import. That sucks.

**Comment by ankurparihar** [comment]
Unable to use glslify with `vite`, please help.

```js
import glslify from 'glslify';

const vertexShader = glslify(`
	varying vec2 vUv;
	uniform float time;

	#pragma glslify: noise = require(glsl-noise/simplex/4d)
	
	void main() {
		vUv = uv;
		vec3 pos = position.xyz;
		pos += noise(vec4(position.xyz, time)) * 1.0;
		gl_Position = projectionMatrix * modelViewMatrix * vec4(pos, 1.0);
	}
`)

<img width="618" alt="Image" src="https://github.com/user-attachments/assets/ae10c75b-b30c-4307-918d-a9a8b0afc112" />

```

**Comment by ericmorand** [comment]
@ankurparihar what error do you get?

**Comment by ankurparihar** [comment]
> [@ankurparihar](https://github.com/ankurparihar) what error do you get?

It seems that glslify is not resolving the pragma shader.

However when I ran glslify from cli (`pnpx glslify test.glsl -o output.glsl`) it worked.

Issue is happening in browser environment.

This is the error I get

<img width="618" alt="Image" src="https://github.com/user-attachments/assets/2b8d8b8e-0352-42c8-b185-e33ea9c7bfcc" />

P.S. If i comment the line 77, my shader is working.

**Comment by ericmorand** [comment]
I tested your code as a standalone node.js module and it fails if I don't install `glsl-noise`. Is it related?

**Comment by ankurparihar** [comment]
> I tested your code as a standalone node.js module and it fails if I don't install `glsl-noise`. Is it related?

Yes, you'll need to install this as well, ` npm i glsl-noise `

**Comment by ankurparihar** [comment]
I've created a codesandbox, there also it's not compiling the shader.

https://codesandbox.io/p/sandbox/ns6yzc

However the error in this case if different. Maybe they're using something else than vite.

**Comment by ericmorand** [comment]
Thanks. Doesn't your error mean that you passed to Three.js the source of your shared instead of the version that is returned  by `glsify`? There is no reason for the pragma to still be there since it is replaced by the content of the noise shader when execution glslify`.

**Comment by ankurparihar** [comment]
Yes, that's what is happening, if I console.log the shader, after passing it to glslify, it remains the same. pragma still remains in it. And that's why three.js is breaking

**Comment by paperschool** [comment]
I am also having this issue in my threejs project - importing glisfy and having it operate on a shader seems to run it thought this method which basically does nothing: 

<img width="457" alt="Image" src="https://github.com/user-attachments/assets/35f5773d-4e5b-4267-ad89-a1f6c21078cb" />
